include/dlaf/matrix/matrix_ref.h

```diff
@@ -0,0 +1,175 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2023, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#pragma once
+
+/// @file
+
+#include <dlaf/matrix/distribution.h>
+#include <dlaf/matrix/matrix.h>
+#include <dlaf/matrix/matrix_base.h>
+#include <dlaf/matrix/tile.h>
+#include <dlaf/types.h>
+
+namespace dlaf::matrix {
+/// A @c MatrixRef represents a sub-matrix of a @c Matrix.
+///
+/// The class has reference semantics, meaning accesses to a @c MatrixRef and
+/// it's corresponding @c Matrix are interleaved if calls to read/readwrite are
+/// interleaved. Access to a @c MatrixRef and its corresponding @c Matrix is not
+/// thread-safe. A @c MatrixRef must outlive its corresponding @c Matrix.
+template <class T, Device D>
+class MatrixRef;
+
```

Comment:
Would this be better named `SubMatrixRef` since that's the primary use case? I'd argue that someone who receives a `MatrixRef` doesn't need to know that it's a sub-matrix of another `Matrix`, but I don't feel too strongly about it.

Reply:
Keeping the name for now until we know better. However, I'm moving this to `internal` until we're more confident with the functionality as suggested by @rasolca.